Protocol update for Table Features (#1450)
delta-io/delta: PROTOCOL.md

      - [Reader Requirements for AddCDCFile](#reader-requirements-for-addcdcfile)
    - [Transaction Identifiers](#transaction-identifiers)
    - [Protocol Evolution](#protocol-evolution)
      - [Feature Struct](#feature-struct)
    - [Commit Provenance Information](#commit-provenance-information)
- [Action Reconciliation](#action-reconciliation)
- [Table Features](#table-features)
- [Column Mapping](#column-mapping)
  - [Writer Requirements for Column Mapping](#writer-requirements-for-column-mapping)
  - [Reader Requirements for Column Mapping](#reader-requirements-for-column-mapping)

> - [Table Features Operations](#table-features-operations)

Reviewer: Why is table features and table feature operations in 2 difference sections?

Reviewer: i think the operations should be under the table features top-level column

#### Writer Requirements for AddCDCFile

As of [Writer Version 4](#Writer-Version-Requirements), all writers must respect the `delta.enableChangeDataFeed` configuration flag in the metadata of the table. When `delta.enableChangeDataFeed` is `true`, writers must produce the relevant `AddCDCFile`'s for any operation that changes data, as specified in [Change Data Files](#change-data-files).

> As of Writer Version 7, all writers must first ensure the feature `changeDataFeed` exists in the table `protocol`'s `writerFeatures` with `status = enabled`, then respect all requirements for Writer Version 4.

Reviewer: Cross link with the table feature section.

The schema of the `protocol` action is as follows:

Field Name | Data Type | Description
-|-|-
minReaderVersion | Int | The minimum version of the Delta read protocol that a client must implement in order to correctly *read* this table
minWriterVersion | Int | The minimum version of the Delta write protocol that a client must implement in order to correctly *write* this table
readerFeatures | Array[Feature Struct] | A collection of features that a client must implement in order to correctly read this table (exist only when `minReaderVersion` is set to `3`)
writerFeatures | Array[Feature Struct] | A collection of features that a client must implement in order to correctly write this table (exist only when `minWriterVersion` is set to `7`)

#### Feature Struct

Field Name | Data Type | Description
-|-|-
name | String | A globally unique _feature name_

> status | String | Indicates whether the table currently uses the feature. The only valid value is `enabled`.

Reviewer 2: what's the point of having a **required** field with only 1 allowed value?
e.g. if just `name` is present, then that could imply that the feature is enabled.

Author: Thanks for the feedback! It does make sense to remove this `status` as it's too hard to add another status with proper backward compatibility (not breaking checkpoints etc.). We just merged https://github.com/delta-io/delta/commit/5c2e6d040092969898f1881c7d95ec278d0b108d to address this suggestion.

Since breaking changes must be accompanied by an increase in the protocol version recorded in a table, clients can assume that unrecognized fields or actions are never required in order to correctly interpret the transaction log.

> 

Reviewer: or do we not want to say that to keep it open-ended still?

> 

Author: Yes, we should mention (3,7) here.

 - To replay the log, scan all file actions and keep only the newest reference for each logical file.
 - `add` actions in the result identify logical files currently present in the table (for queries). `remove` actions in the result identify tombstones of logical files no longer present in the table (for VACUUM).

> # Table Features

Reviewer 3: Strong recommendation: we should remove title casing from all mentions of Table Features.

Even within this initial commit, we have 2 sections where we drop title casing and refer to "enabled features" and "disabled features", and then provide instructions relative to `readerFeatures` and `writerFeatures` rather than using the Table Feature moniker.

We should treat table features the same way we treat protocol version and remove capitalization. While this is an exciting new development for how Delta tables will be versioned today, this will immediately become a commonplace part of the Delta protocol and so not really a "feature" that should need strong emphasis.

Author: Agree, table features will become an essential part of Delta therefore no need to be emphasized.

## Disabled Features
A feature is `disabled` if it is in neither `readerFeatures` nor `writerFeatures`. Writers are allowed to `enable` a feature for the table by adding its name to the `readerFeatures` or `writerFeatures`. Reader features should be added to both `readerFeatures` and `writerFeatures` simultaneously, while writer features should be added only to `writerFeatures`. It is not allowed to add features only to `readerFeatures` but not to `writerFeatures`.

# Column Mapping
Delta can use column mapping to avoid any column naming restrictions, and to support the renaming and dropping of columns without having to rewrite all the data. There are two modes of column mapping, by `name` and by `id`. In both modes, every column - nested or leaf - is assigned a unique _physical_ name, and a unique 32-bit integer as an id. The physical name is stored as part of the column metadata with the key `delta.columnMapping.physicalName`. The column id is stored within the metadata with the key `delta.columnMapping.id`.

> Up to Reader Version 2 and Writer Version 6, the column mapping is governed by the table property `delta.columnMapping.mode` and can be one of `none`, `id`, and `name`. From Reader Version 3 and Writer Version 7, Column Mapping's on/off status is governed by the existence of a feature name `columnMapping` in both `readerFeatures` and `writerFeatures`, plus the table property `delta.columnMapping.mode` being one of `none`, `id` and `name`. In case of Reader version 2 and Writer Version 7 (a table that supports only writer features), Column Mapping's on/off status for readers and writers are governed separately: table property `delta.columnMapping.mode` for readers, and feature name `columnMapping` plus the table property for writers.

Reviewer 4: This talks about protocol versions but it doesn't cover the minimum protocol versions. I.e., "up to writer version 6" doesn't say that it's from writer version 5.

I would suggest separating out the concerns:
1. It's governed by the table property.
2. But only if the feature is enabled in the protocol.

E.g.:
```suggestion
The column mapping is governed by the table property `delta.columnMapping.mode` and can be one of `none`, `id`, and `name`. The table property should only be honored if the table's protocol has the reader/writer versions and/or table features that support the "columnMapping" table feature. For readers this is reader version 2, or reader version 3 with the "columnMapping" table feature enabled. For writers this is writer version 5 or 6, or writer version 7 with  the "columnMapping" table feature enabled.
```

Readers and writers must not ignore Table Features when they are present:
 - to read a table, readers must implement and respect all features listed in `readerFeatures`;
 - to write a table, writers must implement and respect all features listed in `writerFeatures`. Because writers have to read the table (or only the Delta log) before write, they must implement and respect reader features as well.

## Table Features for new and existing tables
It is possible to enable Table Features during table creation or on existing tables. The enablement can be only for readers or both readers and writers.  When a new table is created with a Reader Version up to 2 and Writer Version 7, its `protocol` action must only contain `writerFeatures`. When a new table is created with Reader Version 3 and Writer Version 7, its `protocol` action must contain both `readerFeatures` and `writerFeatures`. Creating a table with a Reader Version 3 and Writer Version up to 6 is not allowed.

> When upgrading an existing table to Reader Version 3 and/or Writer Version 7, the client should, on a best effort basis, determine which features supported by the original protocol version are used in any historical version of the table, and add only used features to Table Features. The client must assume a feature has been used, unless it can prove that the feature is *definitely* not used in any historical version of the table that is reachable by time travel.

Reviewer 2: > the client should, on a best effort basis, determine which features supported by the original protocol version are used in any historical version of the table, and add only used features to Table Features.

This might be a little abstract for readers. Perhaps we could give an example?

e.g. what does it mean or look like when a feature in the original protocol is enabled but not used?

Author: Yep, example is added!

## Enabled features

> A feature is `enabled` when its name is in the `protocol` action’s `readerFeatures` and/or `writerFeatures`. Subsequent read and/or write operations on this table must respect the feature. Clients must not remove the feature from the `protocol` action.

Reviewer 2: this makes me wonder how useful the `status: enabled` key/value is.

Author: We conclude it's not useful because it will be hard to support a new status without another major protocol update. Therefore I will create a PR to remove this `status` from the codebase and update this doc.

## Enabled features
A feature is `enabled` when its name is in the `protocol` action’s `readerFeatures` and/or `writerFeatures`. Subsequent read and/or write operations on this table must respect the feature. Clients must not remove the feature from the `protocol` action.

A feature being `enabled` does not imply it's active, for example, a table may have the [Append-only Tables](#append-only-tables) feature (feature name `appendOnly`) enabled in `writerFeatures`, and have no table property `delta.appendOnly` set to `true`. In such a case the table is not append-only, and writers are allowed to change, remove, and rearrange data. However, writers must implement the feature to know that the table property `delta.appendOnly` should be checked.

## Disabled features
A feature is `disabled` if it’s in neither `readerFeatures` nor `writerFeatures`. Writers are allowed to `enable` this feature for the table, by adding its name to the `protocol` action.

> ## Allowed features

Author: @bart-samwel @ryan-johnson-databricks A starter solution for allowlist ^^

 - To replay the log, scan all file actions and keep only the newest reference for each logical file.
 - `add` actions in the result identify logical files currently present in the table (for queries). `remove` actions in the result identify tombstones of logical files no longer present in the table (for VACUUM).

# Table Features
Table Features must only exist on tables that have a supported protocol version. When the table's Reader Version is 3, `readerFeatures` must exist in the `protocol` action, and when the Writer Version is 7, `writerFeatures` must exist in the `protocol` action. `readerFeatures` and `writerFeatures` define the features that readers and writers must implement in order to read and write this table.

Readers and writers must not ignore Table Features when it’s present:
- to read a table, readers must implement and respect all features listed in `readerFeatures`that have `status = enabled`; readers which support features that have `status = blocked` must not enable those features, while readers which do not support blocked features may still read the table;

> - to write a table, writers must implement and respect all features listed in `writerFeatures` that have `status = enabled`; writers which support features that have `status = blocked` must not enable those features, while writers which do not support blocked features may still write to the table.

Reviewer: I think we should make it an explicit point that effectively all writers have to read the table log, so must satisfy readerfeatures as well.

## Disabled features
A feature is `disabled` if it is in neither `readerFeatures` nor `writerFeatures`. Writers are allowed to `enable` a feature for the table by adding its name to the `readerFeatures` or `writerFeatures`. Reader features should be added to both `readerFeatures` and `writerFeatures` simultaneously, while writer features should be added only to `writerFeatures`. It is not allowed to add features only to `readerFeatures` but not to `writerFeatures`.

# Column Mapping
Delta can use column mapping to avoid any column naming restrictions, and to support the renaming and dropping of columns without having to rewrite all the data. There are two modes of column mapping, by `name` and by `id`. In both modes, every column - nested or leaf - is assigned a unique _physical_ name, and a unique 32-bit integer as an id. The physical name is stored as part of the column metadata with the key `delta.columnMapping.physicalName`. The column id is stored within the metadata with the key `delta.columnMapping.id`.

> Up to Reader Version 2 and Writer Version 6, the column mapping is governed by the table property `delta.columnMapping.mode` and can be one of `none`, `id`, and `name`. From Reader Version 3 and Writer Version 7, Column Mapping's on/off status is governed by the existence of a feature name `columnMapping` in both `readerFeatures` and `writerFeatures`, plus the table property `delta.columnMapping.mode` being one of `none`, `id` and `name`. In case of Reader version 2 and Writer Version 7 (a table that supports only writer Table Features), Column Mapping's on/off status for readers and writers are governed separately: table property `delta.columnMapping.mode` for readers, and feature name `columnMapping` plus table property for writers.

Reviewer 4: I think we can split the table property from the protocol aspects -- the table property is highly repetitive. We just need to know when the protocol supports Column Mapping, and then separately that the table property should only be accessed if the protocol supports column mapping.

## Writer Requirements for Column Mapping
In order to support column mapping, writers must:
 - Write `protocol` and `metaData` actions when Column Mapping is turned on for the first time:

>    - If the table is on a Writer Version up to 6: write a `metaData` action to add the `delta.columnMapping.mode` table property;

Reviewer 4: The writer version must also be set to at least version 5.

 - Data files MUST be uniquely named and MUST NOT be overwritten. The reference implementation uses a GUID in the name to ensure this property.

## Append-only Tables
Enablement:

>  - If the table is on a Writer Version starting from 2 up to 6, the table property `delta.appendOnly` must be set to `true`.
>  - If the table is on Writer Version 7, the feature `appendOnly` must exist in the table `protocol`'s `writerFeatures`, and the table property `delta.appendOnly` must be set to `true`.

Reviewer 4: Can we factor this differently, to separate out the table property which is now duplicated? E.g.:

```suggestion
 - The table must be on a Writer Version between 2 and 7.
 - If the table is on Writer Version 7, the feature `appendOnly` must exist in the table `protocol`'s `writerFeatures`.
- The table property `delta.appendOnly` must be set to `true`.
```

 - Data files MUST be uniquely named and MUST NOT be overwritten. The reference implementation uses a GUID in the name to ensure this property.

## Append-only Tables
Enablement:
 - If the table is on a Writer Version up to 6, the table property `delta.appendOnly` must be set to `true`.
 - If the table is on Writer Version 7, the feature `appendOnly` must exist in the table `protocol`'s `writerFeatures`, and the table property `delta.appendOnly` must be set to `true`.

When enabled:
 - New log entries MUST NOT change or remove data from the table.
 - New log entries may rearrange data (i.e. `add` and `remove` actions where `dataChange=false`).

To remove the append-only restriction, the table property `delta.appendOnly` must be set to `false`.

## Column Invariants
Enablement:

>  - If the table is on a Writer Version up to 6, Column Invariants is enabled when column metadata contains required keys.

Reviewer 4: I would simply say that it is always enabled. I.e., you always have to check the metadata.

Also, doesn't this feature have a _minimum_ protocol version where it was introduced?

## Generated Columns

Enablement:
 - If the table is on a Writer Version starting from 4 up to 6, Generated Columns are always enabled.
 - If the table is on Writer Version 7, a feature name `generatedColumns` must exist in the table `protocol`'s `writerFeatures`.

When enabled:
 - The `metadata` for a column in the table schema MAY contain the key `delta.generationExpression`.
 - The value of `delta.generationExpression` SHOULD be parsed as a SQL expression.
 - Writers MUST enforce that any data writing to the table satisfy the condition `(<value> <=> <generation expression>) IS TRUE`. `<=>` is the NULL-safe equal operator which performs an equality comparison like the `=` operator but returns `TRUE` rather than NULL if both operands are `NULL`

## Identity Columns

Delta supports defining Identity columns on Delta tables. Delta will generate unique values for Identity columns when users do not explicitly provide values for them when writing to such tables. To enable Identity Columns:
 - The table must be on Writer Version 6, or

>  - The table is on Writer Version 7, and a feature name `identityColumns` does exist in the table `protocol`'s `writerFeatures`.

Reviewer 4: ```suggestion
 - The table must be on Writer Version 7, and a feature name `identityColumns` must exist in the table `protocol`'s `writerFeatures`.
```

## CHECK Constraints

Enablement:

> - If the table is on a Writer Version up to 6, CHECK Constraints is enabled when the table's metadata contains required keys.

Reviewer 4: Same: I would suggest saying that it's also enabled. And what is the minimum writer version?

> - If the table is on Writer Version 7, a feature name `checkConstraints` must exist in the table `protocol`'s `writerFeatures`.

Reviewer 4: The feature names are inconsistent in plural / singular. E.g. `checkConstraints` and `deletionVector` and `columnInvariants`. Let's go for plural?

Author: Yep, let's go for plural.

CHECK constraints are stored in the map of the `configuration` field in [Metadata](#change-metadata). Each CHECK constraint has a name and is stored as a key value pair. The key format is `delta.constraints.{name}`, and the value is a SQL expression string whose return type must be `Boolean`. Columns referred by the SQL expression must exist in the table schema.

Rows in a table must satisfy CHECK constraints. In other words, evaluating the SQL expressions of CHECK constraints must return `true` for each row in a table.

For example, a key value pair (`delta.constraints.birthDateCheck`, `birthDate > '1900-01-01'`) means there is a CHECK constraint called `birthDateCheck` in the table and the value of the `birthDate` column in each row must be greater than `1900-01-01`.

Hence, a writer must follow the rules below:

> - Upon the first time a writer add CHECK Constraints to a table and if the table is on Writer Version 7, the writer must write a `protocol` action to add a feature name `checkConstraints` to `writerFeatures`.

Reviewer 4: I would just say that the writer cannot do this unless the feature is enabled (according to the enablement rules above). And that the writer may enable the feature in the same commit where they make use of the feature for the first time.

Make sure this message is consistent across all of the features.

- When adding a CHECK constraint to a table, a writer must validate the existing data in the table and ensure every row satisfies the new CHECK constraint before committing the change. Otherwise, the write must fail and the table must stay unchanged.
- When writing to a table that contains CHECK constraints, every new row being written to the table must satisfy CHECK constraints in the table. Otherwise, the write must fail and the table must stay unchanged.

## Generated Columns

Enablement:

>  - If the table is on a Writer Version up to 6, Generated Columns is enabled when column metadata contains required keys.

Reviewer 4: Same as other cases -- always enabled, and what is the minimum version?

 - If the table is on Writer Version 7, a feature name `generatedColumns` must exist in the table `protocol`'s `writerFeatures`.

When enabled:
 - The `metadata` for a column in the table schema MAY contain the key `delta.generationExpression`.
 - The value of `delta.generationExpression` SHOULD be parsed as a SQL expression.
 - Writers MUST enforce that any data writing to the table satisfy the condition `(<value> <=> <generation expression>) IS TRUE`. `<=>` is the NULL-safe equal operator which performs an equality comparison like the `=` operator but returns `TRUE` rather than NULL if both operands are `NULL`

## Identity Columns

Delta supports defining Identity columns on Delta tables. Delta will generate unique values for Identity columns when users do not explicitly provide values for them when writing to such tables. To enable Identity Columns:

>  - If the table is on a Writer Version up to 6, add required Identity column properties to column metadata.

Reviewer 4: Always enabled, minimum version?

For example, a key value pair (`delta.constraints.birthDateCheck`, `birthDate > '1900-01-01'`) means there is a CHECK constraint called `birthDateCheck` in the table and the value of the `birthDate` column in each row must be greater than `1900-01-01`.

Hence, a writer must follow the rules below:

> - When adding a CHECK Constraint to a table for the first time, writers must submit a `protocol` change in the same commit so that the above enablement rules are satisfied.

Reviewer 4: ```suggestion
- CHECK Constraints may not be added to a table unless the above enablement rules are satisfied. When adding a CHECK Constraint to a table for the first time, writers are allowed to submit a `protocol` change in the same commit to enable the feature in the protocol. 
```

<br> | Requirements
-|-
Writer Version 2 | - Support [Append-only Tables](#append-only-tables)<br>- Support [Column Invariants](#column-invariants)
Writer Version 3 | Enforce:<br>- `delta.checkpoint.writeStatsAsJson`<br>- `delta.checkpoint.writeStatsAsStruct`<br>- [`CHECK` constraints](#check-constraints)
Writer Version 4 | - Support [Change Data Feed](#add-cdc-file)<br>- Support [Generated Columns](#generated-columns)
Writer Version 5 | Respect [Column Mapping](#column-mapping)
Writer Version 6 | Support [Identity Columns](#identity-columns)

> Writer Version 7 | Support [Table Features](#table-features) for writers

Reviewer 4: This list is super inconsistent. Enforce, Support, Respect? Maybe we should standardize on "Respect".

```suggestion
Writer Version 7 | Respect [Table Features](#table-features) for writers
```

<br> | Requirements
-|-
Reader Version 2 | Respect [Column Mapping](#column-mapping)

> Reader Version 3 | Support [Table Features](#table-features) for readers<br> - Writer Version must be 7

Reviewer 4: ```suggestion
Reader Version 3 | Respect [Table Features](#table-features) for readers<br> - Writer Version must be 7
```

# Appendix

## Valid Feature Names in Table Features

> Feature | Name | Required by Readers | Required by Writers
> -|-|-|-
> [Append-only Tables](#append-only-tables) | `appendOnly` | No | Yes
> [Column Invariants](#column-invariants) | `invariants` | No | Yes
> [`CHECK` constraints](#check-constraints) | `checkConstraints` | No | Yes
> [Generated Columns](#generated-columns) | `generatedColumns` | No | Yes
> [Column Mapping](#column-mapping) | `columnMapping` | Yes | Yes
> [Identity Columns](#identity-columns) | `identityColumns` | No | Yes
> [Deletion Vectors](#deletion-vectors) | `deletionVectors` | Yes | Yes

Reviewer 4: ```suggestion
Feature | Name | Readers or Writers?
-|-|-|-
[Append-only Tables](#append-only-tables) | `appendOnly` | Writers only
[Column Invariants](#column-invariants) | `invariants` | Writers only
[`CHECK` constraints](#check-constraints) | `checkConstraints` | Writers only
[Generated Columns](#generated-columns) | `generatedColumns` | Writers only
[Column Mapping](#column-mapping) | `columnMapping` | Readers and writers
[Identity Columns](#identity-columns) | `identityColumns` | Writers only
[Deletion Vectors](#deletion-vectors) | `deletionVectors` | Readers and writers
```

|    |-- dataChange: boolean
```

> If the table is on Reader Version 2 and Writer Version 7, its checkpoint schema will have one additional fields in the protocol struct compared to the previous example:

Reviewer 4: ```suggestion
If the table is on Reader Version 2 and Writer Version 7, its checkpoint schema will have one additional field in the protocol struct compared to the previous example:
```

|    |-- dataChange: boolean
```

> If the table is on Reader Version 3 and Writer Version 7, its checkpoint schema will have two additional fields in the protocol struct compared to the previous example:

Reviewer 4: Also cover Reader Version 1 or 2 with Writer version 7?

|    |-- dataChange: boolean
```

> Please note, as in the above example, `readerFeatures` and `writerFeatures` does exist in the schema even when the table does not support table features. In such a case values of these two fields are `null`.

Reviewer 4: ```suggestion
Please note, as in the above example, the `readerFeatures` and `writerFeatures` fields do exist in the schema even when the table does not support table features. In such a case values of these two fields are `null`.
```
